File: apps/android/app/src/main/java/com/remora/android/state/TerminalSessionController.kt
package com.remora.android.state

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import uniffi.codex_mobile_client.AppStoreInterface
import uniffi.codex_mobile_client.TerminalBackendKind
import uniffi.codex_mobile_client.TerminalOutputEventListener
import uniffi.codex_mobile_client.TerminalOutputSnapshot
import uniffi.codex_mobile_client.TerminalOutputStreamEvent
import uniffi.codex_mobile_client.TerminalOutputSubscription
import uniffi.codex_mobile_client.TerminalSession
import uniffi.codex_mobile_client.TerminalSize
import uniffi.codex_mobile_client.TerminalSshTrustStore

sealed interface TerminalRenderUpdate {
    data class Replace(val bytes: ByteArray) : TerminalRenderUpdate

    data class Append(val bytes: ByteArray) : TerminalRenderUpdate
}

class TerminalSessionController(
    private val scope: CoroutineScope,
    private val appStore: AppStoreInterface = AppModel.shared.store,
) {
    enum class Phase {
        IDLE,
        CONNECTING,
        RUNNING,
        EXITED,
        FAILED,
    }

    data class SshHostTrustChallenge(
        val host: String,
        val port: UShort,
        val fingerprint: String,
        val backend: TerminalBackendKind,
    )

    var phase by mutableStateOf(Phase.IDLE)
        private set
    var output by mutableStateOf("")
        private set
    var exitCode by mutableStateOf<Int?>(null)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set
    var sshTrustChallenge by mutableStateOf<SshHostTrustChallenge?>(null)
        private set

    var sessionId: String? = null
        private set
    private var listener: TerminalOutputEventListener? = null
    private var outputSubscription: TerminalOutputSubscription? = null
    @Volatile
    private var outputSink: ((TerminalRenderUpdate) -> Unit)? = null
    private var outputBytes = ByteArray(0)
    private var expectedSequence: ULong? = null
    private val eventLock = Any()
    private val pendingEvents = ArrayDeque<TerminalOutputStreamEvent>()
    private var eventDrainScheduled = false
    @Volatile
    private var eventGeneration: Int = 0
    private var terminalCols: UShort = 80u
    private var terminalRows: UShort = 24u

    private fun activeSession(): TerminalSession? =
        sessionId?.let { appStore.terminalSessionHandle(it) }

    val canSendInput: Boolean
        get() = phase == Phase.RUNNING

    fun open(backend: TerminalBackendKind) {
        if (sessionId != null || phase == Phase.CONNECTING) return
        eventGeneration += 1
        val generation = eventGeneration
        phase = Phase.CONNECTING
        errorMessage = null
        exitCode = null
        sshTrustChallenge = null
        scope.launch {
            try {
                val size = TerminalSize(cols = terminalCols, rows = terminalRows)
                val id = if (backend is TerminalBackendKind.RemoteSsh) {
                    val backendImpl = SshTrustStore(AppModel.shared.appContext)
                    val trustStore = TerminalSshTrustStore(backendImpl)
                    appStore.openTerminalSessionWithTrustStore(backend, size, trustStore)
                } else {
                    appStore.openTerminalSession(backend, size)
                }
                if (generation != eventGeneration) {
                    runCatching { appStore.closeTerminalSession(id) }
                    return@launch
                }
                val opened = appStore.terminalSessionHandle(id) ?: run {
                    runCatching { appStore.closeTerminalSession(id) }
                    if (generation != eventGeneration) return@launch
                    errorMessage = "Session disappeared after open"
                    phase = Phase.FAILED
                    return@launch
                }
                sessionId = id
                appStore.setActiveTerminalId(id)
                val outputListener = object : TerminalOutputEventListener {
                    override fun onEvent(event: TerminalOutputStreamEvent) {
                        var scheduleDrain = false
                        synchronized(eventLock) {
                            if (generation == eventGeneration) {
                                pendingEvents.addLast(event)
                                if (!eventDrainScheduled) {
                                    eventDrainScheduled = true
                                    scheduleDrain = true
                                }
                            }
                        }
                        if (scheduleDrain) {
                            scope.launch(Dispatchers.Main.immediate) {
                                drainOutputEvents(generation)
                            }
                        }
                    }
                }
                outputSubscription = opened.subscribeOutputEvents(outputListener)
                listener = outputListener
                phase = Phase.RUNNING
            } catch (error: Exception) {
                if (generation != eventGeneration) return@launch
                sessionId = null
                val challenge = sshHostTrustChallenge(error, backend)
                if (challenge != null) {
                    sshTrustChallenge = challenge
                    errorMessage = "Unknown SSH host key ${challenge.fingerprint}"
                } else {
                    errorMessage = error.message ?: "Unable to open terminal"
                }
                phase = Phase.FAILED
            }
        }
    }

    fun trustUnknownSshHostAndRetry() {
        val challenge = sshTrustChallenge ?: return
        SshTrustStore(AppModel.shared.appContext).write(
            host = challenge.host,
            port = challenge.port,
            fingerprint = challenge.fingerprint,
        )
        sshTrustChallenge = null
        errorMessage = null
        phase = Phase.IDLE
        open(challenge.backend)
    }

    fun switchBackend(backend: TerminalBackendKind) {
        close()
        replaceOutput(byteArrayOf())
        open(backend)
    }

    fun send(value: String) {
        sendBytes(value.toByteArray(Charsets.UTF_8))
    }

    fun sendBytes(bytes: ByteArray) {
        if (bytes.isEmpty()) return
        val activeSession = activeSession() ?: return
        if (!canSendInput) return
        scope.launch {
            try {
                activeSession.writeInput(bytes)
            } catch (error: Exception) {
                errorMessage = error.message ?: "Unable to write terminal input"
                phase = Phase.FAILED
            }
        }
    }

    fun sendLine(value: String) {
        send("$value\n")
    }

    fun clearOutput() {
        replaceOutput(byteArrayOf())
    }

    fun setOutputSink(sink: ((TerminalRenderUpdate) -> Unit)?) {
        outputSink = sink
        if (sink == null) {
            output = outputBytes.toString(Charsets.UTF_8)
        }
        sink?.invoke(TerminalRenderUpdate.Replace(outputBytes.copyOf()))
    }

    fun replayOutputToSink() {
        outputSink?.invoke(TerminalRenderUpdate.Replace(outputBytes.copyOf()))
    }

    private fun sshHostTrustChallenge(
        error: Exception,
        backend: TerminalBackendKind,
    ): SshHostTrustChallenge? {
        val sshBackend = backend as? TerminalBackendKind.RemoteSsh ?: return null
        val fingerprint = unknownHostFingerprint(error.message.orEmpty()) ?: return null
        return SshHostTrustChallenge(
            host = sshBackend.host,
            port = sshBackend.port,
            fingerprint = fingerprint,
            backend = backend,
        )
    }

    private fun unknownHostFingerprint(message: String): String? {
        val marker = "unknown-host:"
        val start = message.indexOf(marker)
        if (start < 0) return null
        return message
            .substring(start + marker.length)
            .trim()
            .trim('"', '\'', '(', ')', '[', ']')
            .takeIf { it.isNotEmpty() }
    }

    fun resize(cols: Int, rows: Int, notifyBackend: Boolean = true) {
        if (cols <= 0 || rows <= 0) return
        terminalCols = cols.coerceIn(1, UShort.MAX_VALUE.toInt()).toUShort()
        terminalRows = rows.coerceIn(1, UShort.MAX_VALUE.toInt()).toUShort()
        val activeSession = activeSession() ?: return
        if (!notifyBackend || !canSendInput) return
        val size = TerminalSize(cols = terminalCols, rows = terminalRows)
        scope.launch {
            try {
                activeSession.resize(size)
            } catch (error: Exception) {
                errorMessage = error.message ?: "Unable to resize terminal"
                phase = Phase.FAILED
            }
        }
    }

    fun close() {
        eventGeneration += 1
        outputSubscription?.cancel()
        outputSubscription = null
        listener = null
        synchronized(eventLock) {
            pendingEvents.clear()
            eventDrainScheduled = false
        }
        val id = sessionId
        sessionId = null
        phase = Phase.IDLE
        if (id == null) return
        scope.launch {
            runCatching { appStore.closeTerminalSession(id) }
        }
    }

    private fun drainOutputEvents(generation: Int) {
        val events = synchronized(eventLock) {
            val drained = pendingEvents.toList()
            pendingEvents.clear()
            eventDrainScheduled = false
            drained
        }
        if (generation != eventGeneration) return
        events.forEach(::applyOutputEvent)
    }

    private fun applyOutputEvent(event: TerminalOutputStreamEvent) {
        when (event) {
            is TerminalOutputStreamEvent.Snapshot -> applySnapshot(event.snapshot)
            is TerminalOutputStreamEvent.Reset -> applySnapshot(event.snapshot)
            is TerminalOutputStreamEvent.Output -> applyOutput(event.sequence, event.data)
            is TerminalOutputStreamEvent.Exited -> {
                val expected = expectedSequence
                if (expected == null || event.sequence >= expected) {
                    expectedSequence = event.sequence + 1u
                    exitCode = event.code
                    phase = Phase.EXITED
                }
            }
        }
    }

    private fun applySnapshot(snapshot: TerminalOutputSnapshot) {
        expectedSequence = snapshot.latestSequence?.plus(1u) ?: snapshot.baseSequence
        replaceOutput(snapshot.bytes)
        snapshot.exitCode?.let { code ->
            exitCode = code
            phase = Phase.EXITED
        }
    }

    private fun applyOutput(sequence: ULong, data: ByteArray) {
        val expected = expectedSequence
        if (expected != null) {
            if (sequence < expected) return
            if (sequence > expected) {
                activeSession()?.let { applySnapshot(it.outputSnapshot()) }
                return
            }
        }
        expectedSequence = sequence + 1u
        appendOutput(data)
    }

    private fun replaceOutput(data: ByteArray) {
        outputBytes = boundedOutput(data)
        output = outputBytes.toString(Charsets.UTF_8)
        outputSink?.invoke(TerminalRenderUpdate.Replace(outputBytes.copyOf()))
    }

    private fun appendOutput(data: ByteArray) {
        if (data.isEmpty()) return
        outputBytes = boundedOutput(outputBytes, data)
        if (outputSink == null) {
            output = outputBytes.toString(Charsets.UTF_8)
        }
        outputSink?.invoke(TerminalRenderUpdate.Append(data.copyOf()))
    }

    private fun boundedOutput(data: ByteArray): ByteArray {
        val limit = 64 * 1024
        if (data.size <= limit) return data.copyOf()
        return data.copyOfRange(data.size - limit, data.size)
    }

    private fun boundedOutput(existing: ByteArray, appended: ByteArray): ByteArray {
        val limit = 64 * 1024
        if (appended.size >= limit) {
            return appended.copyOfRange(appended.size - limit, appended.size)
        }
        val existingCount = minOf(existing.size, limit - appended.size)
        return ByteArray(existingCount + appended.size).also { combined ->
            existing.copyInto(
                destination = combined,
                destinationOffset = 0,
                startIndex = existing.size - existingCount,
            )
            appended.copyInto(combined, destinationOffset = existingCount)
        }
    }
}

File: apps/ios/Sources/Remora/Models/TerminalSessionController.swift
import Foundation
import Observation

enum TerminalRenderUpdate: Sendable {
    case replace(Data)
    case append(Data)
}

@MainActor
@Observable
final class TerminalSessionController {
    enum Phase: Equatable {
        case idle
        case connecting
        case running
        case exited(Int32)
        case failed(String)
    }

    struct SshHostTrustChallenge {
        let host: String
        let port: UInt16
        let fingerprint: String
        let backend: TerminalBackendKind
    }

    private(set) var phase: Phase = .idle
    private(set) var output = ""
    private(set) var sessionId: String?
    private(set) var sshTrustChallenge: SshHostTrustChallenge?

    @ObservationIgnored private let appStore: AppStore
    @ObservationIgnored private var outputListener: TerminalOutputRelay?
    @ObservationIgnored private var outputSubscription: TerminalOutputSubscription?
    @ObservationIgnored private var outputSink: ((TerminalRenderUpdate) -> Void)?
    @ObservationIgnored private var outputBytes = Data()
    @ObservationIgnored private var expectedSequence: UInt64?
    @ObservationIgnored private var eventGeneration = 0
    @ObservationIgnored private var terminalSize = TerminalSize(cols: 80, rows: 24)

    init() {
        self.appStore = AppModel.shared.store
    }

    init(appStore: AppStore) {
        self.appStore = appStore
    }

    var canSendInput: Bool {
        if case .running = phase { return true }
        return false
    }

    func open(backend: TerminalBackendKind) async {
        guard sessionId == nil else { return }
        eventGeneration &+= 1
        let generation = eventGeneration
        phase = .connecting
        sshTrustChallenge = nil
        do {
            let id: String
            if isSshBackend(backend) {
                let trustStore = TerminalSshTrustStore(backend: SwiftSshTrustBackend.shared)
                id = try await appStore.openTerminalSessionWithTrustStore(
                    kind: backend,
                    size: terminalSize,
                    trustStore: trustStore
                )
            } else {
                id = try await appStore.openTerminalSession(
                    kind: backend,
                    size: terminalSize
                )
            }
            guard generation == eventGeneration else {
                try? await appStore.closeTerminalSession(id: id)
                return
            }
            guard let session = appStore.terminalSessionHandle(id: id) else {
                try? await appStore.closeTerminalSession(id: id)
                guard generation == eventGeneration else { return }
                phase = .failed("Session disappeared after open")
                return
            }
            sessionId = id
            appStore.setActiveTerminalId(id: id)
            let listener = TerminalOutputRelay(owner: self, generation: generation)
            outputSubscription = session.subscribeOutputEvents(listener: listener)
            outputListener = listener
            phase = .running
        } catch {
            guard generation == eventGeneration else { return }
            sessionId = nil
            if let challenge = Self.sshHostTrustChallenge(from: error, backend: backend) {
                sshTrustChallenge = challenge
                phase = .failed("Unknown SSH host key \(challenge.fingerprint)")
            } else {
                phase = .failed(error.localizedDescription)
            }
        }
    }

    private func isSshBackend(_ backend: TerminalBackendKind) -> Bool {
        if case .remoteSsh = backend { return true }
        return false
    }

    func trustUnknownSshHostAndRetry() async {
        guard let challenge = sshTrustChallenge else { return }
        // Record through the Rust store, not the backend directly: `pin`
        // applies the same host normalization (case, brackets, IPv6 zone id)
        // that the connect-time lookup uses. Writing the raw challenge host
        // straight to the backend would file the approval under a
        // noncanonical key, leaving the canonical spelling unpinned and still
        // eligible for trust-on-first-use.
        do {
            try TerminalSshTrustStore(backend: SwiftSshTrustBackend.shared).pin(
                host: challenge.host,
                port: challenge.port,
                fingerprint: challenge.fingerprint
            )
        } catch {
            phase = .failed(error.localizedDescription)
            return
        }
        sshTrustChallenge = nil
        phase = .idle
        await open(backend: challenge.backend)
    }

    func switchBackend(_ backend: TerminalBackendKind) async {
        close()
        replaceOutput(Data())
        await open(backend: backend)
    }

    func send(_ string: String) async {
        await send(Data(string.utf8))
    }

    func send(_ data: Data) async {
        guard let id = sessionId, canSendInput else { return }
        guard let session = appStore.terminalSessionHandle(id: id) else { return }
        do {
            try await session.writeInput(data: data)
        } catch {
            phase = .failed(error.localizedDescription)
        }
    }

    func sendLine(_ string: String) async {
        await send(string + "\n")
    }

    func clearOutput() {
        replaceOutput(Data())
    }

    func setOutputSink(_ sink: ((TerminalRenderUpdate) -> Void)?) {
        outputSink = sink
        if sink == nil {
            output = String(decoding: outputBytes, as: UTF8.self)
        }
        sink?(.replace(outputBytes))
    }

    private static func sshHostTrustChallenge(
        from error: Error,
        backend: TerminalBackendKind
    ) -> SshHostTrustChallenge? {
        guard case .remoteSsh(
            host: _,
            port: _,
            username: _,
            auth: _,
            shell: _,
            acceptUnknownHost: _,
            cwd: _
        ) = backend else {
            return nil
        }
        guard let terminalError = error as? TerminalError else {
            return nil
        }
        guard case let .SshHostKeyVerification(host, port, fingerprint, pinned) = terminalError,
              pinned == nil else { return nil }
        return SshHostTrustChallenge(
            host: host,
            port: port,
            fingerprint: fingerprint,
            backend: backend
        )
    }

    func resize(cols: UInt16, rows: UInt16, notifyBackend: Bool = true) async {
        guard cols > 0, rows > 0 else { return }
        let size = TerminalSize(cols: cols, rows: rows)
        terminalSize = size
        guard notifyBackend, let id = sessionId, canSendInput else { return }
        guard let session = appStore.terminalSessionHandle(id: id) else { return }
        do {
            try await session.resize(size: size)
        } catch {
            phase = .failed(error.localizedDescription)
        }
    }

    func close() {
        eventGeneration &+= 1
        outputSubscription?.cancel()
        outputSubscription = nil
        outputListener?.deactivate()
        outputListener = nil
        let id = sessionId
        sessionId = nil
        phase = .idle
        guard let id else { return }
        if appStore.activeTerminalId() == id {
            appStore.setActiveTerminalId(id: nil)
        }
        Task.detached(priority: .utility) { [appStore] in
            try? await appStore.closeTerminalSession(id: id)
        }
    }

    fileprivate func applyOutputEvents(
        _ events: [TerminalOutputStreamEvent],
        generation: Int
    ) {
        guard generation == eventGeneration else { return }
        for event in events {
            switch event {
            case let .snapshot(snapshot), let .reset(snapshot):
                applySnapshot(snapshot)
            case let .output(sequence, data):
                applyOutput(sequence: sequence, data: data)
            case let .exited(sequence, code):
                if let expectedSequence, sequence < expectedSequence { continue }
                expectedSequence = sequence &+ 1
                phase = .exited(code)
            }
        }
    }

    private func applySnapshot(_ snapshot: TerminalOutputSnapshot) {
        expectedSequence = snapshot.latestSequence.map { $0 &+ 1 } ?? snapshot.baseSequence
        replaceOutput(snapshot.bytes)
        if let exitCode = snapshot.exitCode {
            phase = .exited(exitCode)
        }
    }

    private func applyOutput(sequence: UInt64, data: Data) {
        if let expectedSequence {
            if sequence < expectedSequence { return }
            if sequence > expectedSequence,
               let id = sessionId,
               let session = appStore.terminalSessionHandle(id: id) {
                applySnapshot(session.outputSnapshot())
                return
            }
        }
        expectedSequence = sequence &+ 1
        appendOutput(data)
    }

    private func replaceOutput(_ data: Data) {
        outputBytes = boundedOutput(data)
        output = String(decoding: outputBytes, as: UTF8.self)
        outputSink?(.replace(outputBytes))
    }

    private func appendOutput(_ data: Data) {
        guard !data.isEmpty else { return }
        outputBytes.append(data)
        outputBytes = boundedOutput(outputBytes)
        if outputSink == nil {
            output = String(decoding: outputBytes, as: UTF8.self)
        }
        outputSink?(.append(data))
    }

    private func boundedOutput(_ data: Data) -> Data {
        let maxCount = 64 * 1024
        guard data.count > maxCount else { return data }
        return Data(data.suffix(maxCount))
    }
}

private final class TerminalOutputRelay: TerminalOutputEventListener, @unchecked Sendable {
    private weak var owner: TerminalSessionController?
    private let generation: Int
    private let lock = NSLock()
    private var pendingEvents: [TerminalOutputStreamEvent] = []
    private var drainScheduled = false
    private var active = true

    init(owner: TerminalSessionController, generation: Int) {
        self.owner = owner
        self.generation = generation
    }

    func deactivate() {
        lock.lock()
        active = false
        pendingEvents.removeAll()
        lock.unlock()
    }

    func onEvent(event: TerminalOutputStreamEvent) {
        var shouldSchedule = false
        lock.lock()
        if active {
            pendingEvents.append(event)
            if !drainScheduled {
                drainScheduled = true
                shouldSchedule = true
            }
        }
        lock.unlock()

        if shouldSchedule {
            Task { @MainActor [weak self] in
                self?.drain()
            }
        }
    }

    @MainActor
    private func drain() {
        lock.lock()
        let events = active ? pendingEvents : []
        pendingEvents.removeAll(keepingCapacity: true)
        drainScheduled = false
        lock.unlock()
        guard !events.isEmpty else { return }
        owner?.applyOutputEvents(events, generation: generation)
    }
}
